lib/core/src/Cardano/Wallet/DB/Store/Transactions/Store.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE NoMonomorphismRestriction #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{- |
+Copyright: © 2018-2022 IOHK
+License: Apache-2.0
+
+Implementation of a 'Store' for 'TxHistory'.
+
+-}
+module Cardano.Wallet.DB.Store.Transactions.Store
+    ( selectTxHistory
+    , putTxHistory
+    , mkStoreTransactions ) where
+
+import Prelude
+
+import Cardano.Wallet.DB.Sqlite.Schema
+    ( EntityField (..)
+    , TxCollateral (..)
+    , TxCollateralOut (..)
+    , TxCollateralOutToken (..)
+    , TxIn (..)
+    , TxOut (..)
+    , TxOutToken (..)
+    , TxWithdrawal (..)
+    )
+import Cardano.Wallet.DB.Sqlite.Types
+    ( TxId )
+import Cardano.Wallet.DB.Store.Transactions.Model
+    ( DeltaTxHistory (..)
+    , TxHistory
+    , TxHistoryF (TxHistoryF)
+    , TxRelationF (..)
+    , TxRelationTxInContextK (NoTxInContext)
+    , tokenCollateralOrd
+    , tokenOutOrd
+    )
+import Control.Arrow
+    ( Arrow ((&&&)) )
+import Data.DBVar
+    ( Store (..) )
+import Data.Foldable
+    ( fold, forM_, toList )
+import Data.List
+    ( sortOn )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( maybeToList )
+import Data.Monoid
+    ( getFirst )
+import Database.Persist.Sql
+    ( Entity
+    , SqlPersistT
+    , deleteWhere
+    , entityVal
+    , insertMany_
+    , selectList
+    , (==.)
+    )
+
+import qualified Data.Map.Strict as Map
+
+mkStoreTransactions
+    :: Store (SqlPersistT IO) DeltaTxHistory
+mkStoreTransactions =
+    Store
+    { loadS = Right <$> selectTxHistory
+    , writeS = write
+    , updateS = update
+    }
+
+update :: TxHistory -> DeltaTxHistory -> SqlPersistT IO ()
+update _ change = case change of
+    Expand txs -> putTxHistory txs
```

Comment:
I think that we do have to use `repsertMany` in the definition of `putTxHistory`, otherwise corner cases like `apply (Expand txs1) txs1` will throw an error. 🤔

The corner case is not pretty, but these delta encodings don't work well with partial functions, `apply` should always give a result (a no-op in the worst case).

Reply:
It will obviously do no harm. My point here is that "as transactions do not change" repserting is working is a symonym of broken code that rewrite transactions. I'm not happy with repsert, we have our copy in memory we should decide there what to do on the database and make only necessary changes (I mean defend the persistence layer) . But I see your points.

Reply:
done